### PR TITLE
RFC: [Feature] Non-Success Type Checking - Responses Union Type #73

### PR DESCRIPTION
**Pull request by scottc**
RFC: [Feature] Non-Success Type Checking - Responses Union Type #73

There are breaking API changes, that may impact users of this node module.

### CODE REVIEW

File: __tests__/__snapshots__/runner.js.snap [hunk 1]
```diff
@@ -907,7 +907,7 @@ export class PetshopApi {
         $queryParameters ? : any,
         $domain ? : string,
         $path ? : string | ((path: string) => string)
-    }): Promise < ResponseWithBody < void >> {
+    }): Promise < ResponseWithBody < object >> {
```
Comment (scottc) [inline __tests__/__snapshots__/runner.js.snap]: Potentially a regression.

File: __tests__/__snapshots__/runner.js.snap [hunk 2]
```diff
@@ -976,7 +976,7 @@ export class PetshopApi {
         $queryParameters ? : any,
         $domain ? : string,
         $path ? : string | ((path: string) => string)
-    }): Promise < ResponseWithBody < void >> {
+    }): Promise < ResponseWithBody < object | object | object >> {
```
Comment (scottc) [inline __tests__/__snapshots__/runner.js.snap]: This type signature could be simplified

File: src/view-data/method.ts [hunk 3]
```diff
@@ -28,8 +28,7 @@ export interface Method {
   readonly externalDocs: string;
   readonly parameters: TypeSpecParameter[];
   readonly headers: Header[];
-  readonly successfulResponseType: string;
-  readonly successfulResponseTypeIsRef: boolean;
+  readonly responseTypes: string;
```
Comment (scottc) [inline src/view-data/method.ts]: Breaking API change.

File: __tests__/__snapshots__/runner.js.snap (hunk 2 above)
Comment (mtennoe) [reply __tests__/__snapshots__/runner.js.snap]: Lets try to do that as part of this PR

File: src/view-data/method.ts (hunk 3 above)
Comment (mtennoe) [reply src/view-data/method.ts]: Thinking a bit about this - Let's try to keep it backwards compatible. We can keep the old ones and mark them as deprecated, and remove them at a later point in time

File: src/view-data/responseType.ts [hunk 4]
```diff
@@ -1,58 +1,11 @@
...
-      convertedType.target ||
-      convertedType.tsType ||
-      defaultSuccessfulResponseType;
-  } catch (error) {
-    successfulResponseType = defaultSuccessfulResponseType;
-  }
-
-  return [successfulResponseType, successfulResponseTypeIsRef];
-}
+export const getResponseTypes = (op: HttpOperation, swagger: Swagger): string =>
```
Comment (mtennoe) [inline src/view-data/responseType.ts]: A lot less code, like it!

**Comment by mtennoe** [comment]
Thanks for giving it a go!

**Comment by scottc** [comment]
Updated pull request:
- Keep `successfulResponseType` and `successfulResponseTypeIsRef` backwards compatibility
- Simplify response union type by culling duplicate types

File: __tests__/__snapshots__/runner.js.snap [hunk 5]
```diff
@@ -1611,7 +1611,7 @@ export class PetshopApi {
         $queryParameters ? : any,
         $domain ? : string,
         $path ? : string | ((path: string) => string)
-    }): Promise < ResponseWithBody < void >> {
+    }): Promise < ResponseWithBody < object >> {
```
Comment (mtennoe) [inline __tests__/__snapshots__/runner.js.snap]: Hm, should we attempt to keep these as void (i.e. avoid the object) to show clearer that no data is returned? Semantically that sounds cleaner

Comment (scottc) [reply __tests__/__snapshots__/runner.js.snap]: I agree, `object` type would be incorrect. And a non-value type like `void` would be better suited.

The PetshopApi response types:
```
"responses": { "405": { "description": "Invalid input" } },
```

The spec:
> To indicate the response body is empty, do not specify a content for the response

https://swagger.io/docs/specification/describing-responses/#empty

Any type that could represent an "empty" non-value type would be a good candidate. `void`, `undefined`, `null` etc.

Also for the sake of backwards compatibility `void` is preferred.

I'll try fix this issue tomorrow.

Comment (scottc) [reply __tests__/__snapshots__/runner.js.snap]: Apologies, I've been inactive. An update.

I had a go at implementing the changes, it needs some changes in typescript.ts convertType(...) function.

Needs an explicit void result type conversion.

```
export function convertType(
  swaggerType: SwaggerType,
  swagger: Swagger
): TypeSpec {
  if (isSchema(swaggerType)) {
    return convertType(swaggerType.schema, swagger);
  } else if (isReference(swaggerType)) {
    return makeReferenceTypeSpec(swaggerType);
  } else if (isEnum(swaggerType)) {
    return makeEnumTypeSpec(swaggerType);
  } else if (isString(swaggerType)) {
    return makeStringTypeSpec(swaggerType);
  } else if (isNumber(swaggerType)) {
    return makeNumberTypeSpec(swaggerType);
  } else if (isBoolean(swaggerType)) {
    return makeBooleanTypeSpec(swaggerType);
  } else if (isArray(swaggerType)) {
    return makeArrayTypeSpec(swaggerType, swagger);
  } else if (isDictionary(swaggerType)) {
    // case where a it's a Dictionary<string, someType>
    return makeDictionaryTypeSpec(swaggerType, swagger);
  } else if (
    swaggerType.minItems >= 0 &&
    swaggerType.hasOwnProperty("title") &&
    !swaggerType.$ref
  ) {
    return makeAnyTypeSpec(swaggerType);
  }

  // Remaining types are created as objects
  return makeObjectTypeSpec(swaggerType, swagger);
}
```

I'll have another shot, just I have limited time and effort.
And some other real life priorities.

Comment (mtennoe) [reply __tests__/__snapshots__/runner.js.snap]: No worries, there is no rush, just take your time! :)

File: src/type-mappers/void.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
+import { makeTypeSpecFromSwaggerType, TypeSpec } from "../typespec";
+// import {
```
Comment (mtennoe) [inline src/type-mappers/void.ts]: remove unused imports

File: src/type-mappers/void.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+//   isArray,
+//   uniqBy,
+//   reverse
+// } from "lodash";
+import { SwaggerType } from "../swagger/Swagger";
+//import { Swagger } from "../swagger/Swagger";
+//import { convertType } from "../typescript";
+
+export interface VoidTypeSpec extends TypeSpec {
+  readonly tsType: "void";
```
Comment (mtennoe) [inline src/type-mappers/void.ts]: remove unused props

File: __tests__/__snapshots__/runner.js.snap (hunk 5 above)
Comment (mtennoe) [reply __tests__/__snapshots__/runner.js.snap]: This looks good!

File: src/view-data/responseType.ts [hunk 8]
```diff
@@ -1,9 +1,20 @@
 import { convertType } from "../typescript";
 import { HttpOperation, Swagger, SwaggerType } from "../swagger/Swagger";
+import { values, uniq } from "lodash/fp";
 
-const defaultSuccessfulResponseType = "void";
+const defaultResponseType = "void";
+
+export const getResponseTypes = (op: HttpOperation, swagger: Swagger): string =>
+  uniq(
```
Comment (mtennoe) [inline src/view-data/responseType.ts]: Can we make this slightly more readable with a few comments and helper consts?

**Comment by mtennoe** [comment]
Looks like tests need to be updated

**Comment by scottc** [comment]
> Looks like tests need to be updated

I've updated the snapshots.
But there is 1 unit test still failing.

> FAIL  lib/typescript.test.js
>  ● convertType › array › correctly passes through typespec properties

It's expecting `Array<object>`, but we actually have `Array<void>`, and a few other differences.

Will need to read into the spec for array details.

**Comment by scottc** [comment]
The 3.0 Spec:
https://swagger.io/docs/specification/data-models/data-types/#array

> **Arrays**
> Arrays are defined as:
```
type: array
items:
  type: string
```
> Unlike JSON Schema, the items keyword is required in arrays. The value of items is a schema that describes the type and format of array items.

The test:
```
it("correctly passes through typespec properties", () => {
  swaggerType = makeSwaggerType({
    description: "The description of a array property",
    required: false,
    type: "array",
    items: [] // this seems to be invalid, and should be some kind of valid swagger type.
  });

  expect(convertType(swaggerType, swagger)).toEqual({
    ...emptyTypeSpecWithDefaults,
    description: "The description of a array property",
    isRequired: false,
    isNullable: true,
    isArray: true,
    tsType: "Array<object>",
    isAtomic: false,
    elementType: {
      ...emptyTypeSpecWithDefaults,
      isRequired: false,
      isNullable: true,
      isObject: true,
      properties: [],
      tsType: "object",
      requiredPropertyNames: []
    }
  });
});
```

**Comment by scottc** [comment]
I've added support for http status code information in the union types, in another branch.

See comment:
https://github.com/mtennoe/swagger-typescript-codegen/issues/73#issuecomment-495463134

And branch / commit:
https://github.com/scottc/swagger-typescript-codegen/tree/http-codes
https://github.com/scottc/swagger-typescript-codegen/commit/30fdb817344ea5ecc3120a8463e96f2078203ae3

**Comment by mtennoe** [comment]
Nice, that would be very nice to have. You can merge it in with the PR to keep everything contained (imo)

**Comment by scottc** [comment]
Added to pull request.

Found why tests are failing, generating some types that look like this.

```
}): Promise < ResponseWithBody < 200, Authentication > | ResponseWithBody < 401, Error > | ResponseWithBody < security, void > | ResponseWithBody < default, Error >> {
```

```
"responses": {
    "200": {
        "description": "Token",
        "schema": {
            "$ref": "#/definitions/Authentication"
        }
    },
    "security": [
        {
            "oauth2": []
        }
    ],
    "401": {
        "description": "Unauthorized",
        "schema": {
            "$ref": "#/definitions/Error"
        }
    },
    "default": {
        "description": "Error",
        "schema": {
            "$ref": "#/definitions/Error"
        }
    }
}
```

**Re: default**

>You can use the default response to describe these errors collectively, not individually. “Default” means this response is used for all HTTP codes that are not covered individually for this operation.

https://swagger.io/docs/specification/2-0/describing-responses/#default
https://swagger.io/docs/specification/describing-responses/#default

Seems that these should be converted into a catch all http status code `number` typescript type.

File: src/view-data/method.ts [hunk 9]
```diff
@@ -28,8 +32,13 @@ export interface Method {
   readonly externalDocs: string;
   readonly parameters: TypeSpecParameter[];
   readonly headers: Header[];
+  readonly responseTypes: string;
+
+  /** @deprecated use responseTypes instead, this field will be removed in a future version. */
   readonly successfulResponseType: string;
+  /** @deprecated use responseTypes instead, this field will be removed in a future version. */
   readonly successfulResponseTypeIsRef: boolean;
+  /** @deprecated use responseTypes2 instead */
```
Comment (mtennoe) [inline src/view-data/method.ts]: responseTypes2 -> responseTypes

**Comment by mtennoe** [comment]
Hm, what is the security type in `ResponseWithBody < security, void >`?

**Comment by scottc** [comment]
> Hm, what is the security type in `ResponseWithBody < security, void >`?

3.0 - https://swagger.io/docs/specification/authentication/#security
2.0 - https://swagger.io/docs/specification/2-0/authentication/

> you can apply them to the whole API or individual operations by adding the `security` section on the root level or operation level, respectively. When used on the root level, `security` applies the specified security schemes globally to all API operations, unless overridden on the operation level.

I suspect that the example schema is invalid, seems that `security` should be defined as a child of root or a child of an operation ie `get` or `post`.  Instead of a child of `responses`.

File: src/view-data/method.ts (hunk 9 above)
Comment (scottc) [reply src/view-data/method.ts]: This JSDoc comment was for a field that no longer exists.

I removed the comment.

File: src/view-data/responseType.ts (hunk 8 above)
Comment (scottc) [reply src/view-data/responseType.ts]: Added JSDocs and helper functions

File: __tests__/__snapshots__/runner.js.snap (hunk 2 above)
Comment (scottc) [reply __tests__/__snapshots__/runner.js.snap]: Added Uniq.

**Comment by mtennoe** [comment]
Cool, thanks for the docs!
Looks good now then, merging

**Comment by mtennoe** [comment]
Published as 1.10.0

**Pull request merged**